docs/framework-design.md

```diff
@@ -0,0 +1,108 @@
+
+### Requirements for consuming repositories
+
+Consuming repositories must have the framework repository as a Git submodule for development work and CI scripts. For merely using the content of the consuming repository, in particular building and running tests, it is enough to have the content of the framework repository, for example from a zip file.
+
+Release archives must include the content of the framework repository.
+
+Without the content of the framework repository, there is no expectation that the consuming repository is usable. In particular, we do not expect the following to work:
+
+* Generating configuration-independent files (e.g. `make generated_files`), not even the ones in the `library` directory.
+* `make lib` (with GNU make or CMake) from a pristine checkout.
+* `make test` (even if all the tests have been built).
+
+We expect `make lib` to work if the generated files in `library` are present and the content of the `framework` directory is missing. This means that in release tags, which include the generated configuration-independent files, we expect `make lib` to work.
+
+## Contents of the framework repository
+
+### Criteria for inclusion
+
+In general, a file should be in the framework repository if it is expected to be present with near-identical content in two or more consuming repositories. Some files have a significant proportion of shared content and should be split into a shared part and a non-shared part.
+
+## CI architecture
+
+Tasks:
+
+* CI in consuming repositories must support Git submodules. Other than that, keep the CI as it is now. In particular, the CI in consuming repositories does not need to consider anything but the commit that the framework submodule points to.
+* CI in the framework repository should run a subset of the CI of all consuming repositories, to warn about unintended breakage. This way, most of the time, updating the framework submodule in a consuming repository to the tip of the `main` branch should work. Gatekeepers can bypass this check if the incompatibility is deliberate.
+* When merging a pull request to in an official branch in a consuming repository (`development`, LTS branches), check that the commit.
+
+TODO: once this is set up, detail the processes here.
+
+## How to make a change
+
+### Change in a consuming repository requiring a new framework feature
+
+If a change in a consuming repository requires a new feature in the framework, you need to make both a pull request in the framework repository and a pull request in the framework repository.
+
+1. Make a pull request (PR) in the framework repository.
+2. Upload the framework branch to the framework repository itself (not a fork). This is necessary for the commit to be available on the CI of the consuming repositories (and also for it to be conveniently available to reviewers).
```

Comment:
These should be part of the same step or possibly reversed, IIUC. A new branch of the `framework` repository should be created (in the upstream repository, not a fork) and then a pull request should be raised from this branch against `main`.

Reply:
Not necessarily: for review, the head branch doesn't have to be in the framework repository. That being said, maybe we should make it a policy that in the framework, all pull requests must come from a repository branch? External contributors can't do that though (you need write permission to the repository), which would be a problem sometimes.

This step is only necessary if we want to do CI without more tooling hassle. But we could arrange CI tooling that's capable of using branches from forks (I don't think that's difficult, just not done AFAICT). I think I'm going to leave this section as is, that's an aspect for your subteam to ponder this quarter.